### ISSUE TITLE
Make loading default properties easier and more consistent

### ISSUE THREAD

**Issue by asarkar** [micronaut-projects/micronaut-core#5176]
**Context:**
Currently, properties can be defaulted in `application.*` file, or in the `package-info.java` file as shown below.
```
@PropertySource({@Property(name = "jpa.default.entity-scan.packages", value = "com.mycompany.package")})
package com.mycompany.package;

import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.PropertySource;
```

`@MicronautTest` additionally has an attribute `propertySources` for loading test properties.

**Problems:**

1. Libraries don't have `application.*` files. While it's possible to set defaults in their own `@ConfigurationProperties`, the only way to set default values for external properties (like `jpa.default.entity-scan.packages`) is using `@PropertySource` as shown above, which becomes cumbersome as the number of properties grow.
2. Kotlin libraries have to create a `package-info.java` merely for setting defaults.
3. `@MicronautTest` has yet another way of setting properties.

**Suggestions:**

1. Update `@PropertySource` such that it can accept a list of key value pairs (more succinct that having to specify `name` and `value` separately. Add attribute `resources` or `files` that can load properties from files. For example:
```
@PropertySource(
    properties = ["a=b"],
    files = ["classpath:defaults.yml"]
)
```

2. Remove `propertySources` from `@MicronautTest`, and make `@PropertySource` the only way to set properties in code. `application.*` files can coexist.

**Comment by graemerocher** [comment]
Note you can always define your own `PropertySourceLoader` implementation if you need to load properties from a custom location for a library.

**Comment by asarkar** [comment]
@graemerocher Can you explain what do you mean by "custom location"? Micronaut libraries don't look for `application.yml` on the classpath, do they? If not, then any file is custom, isn't it?

**Comment by graemerocher** [comment]
What I mean is you can implement `PropertySourceLoader` and from the `load(String resourceName, ResourceLoader resourceLoader)` method load whatever resource you want from the passed resource loader from the classpath whether it is from a library or an application

You just need to call `resourceLoader.getResource("classpath:defaults.yml") I believe. Thought it would certainly be better to put that in package of the library to avoid conflicts

**Comment by asarkar** [comment]
Had the team made a decision (based on label `under consideration`)? I certainly believe having one annotation for all property needs would be a good addition to the API.

**Comment by brianwyka** [comment]
@asarkar, I was able to achieve loading configurations from a library in addition to the application by adding a custom `PropertySourceLoader` in the service file...

Example `META-INF/services/io.micronaut.context.env.PropertySourceLoader` file:
```
com.example.CustomPropertySourceLoader
io.micronaut.context.env.yaml.YamlPropertySourceLoader
io.micronaut.context.env.PropertiesPropertySourceLoader
```
Overriding this one: https://github.com/micronaut-projects/micronaut-core/blob/2.4.x/inject/src/main/resources/META-INF/services/io.micronaut.context.env.PropertySourceLoader

In the `CustomPropertySourceLoader` I load an additional `yml` configuration file with a resource name other than `application` and that one will be the default.  Any `application.yml` or `application-env.yml` or `application.properties` will override those defaults...

@graemerocher, perhaps in addition to the default application name `application`, a user can provide a list of other names that (in listed order) provide default or baseline configurations?

**Comment by asarkar** [comment]
>  by adding a custom PropertySourceLoader in the service file

@brianwyka The purpose of this ticket is to provide OOTB support for such boilerplate code.

**Comment by brianwyka** [comment]
@asarkar, yes.  Was just providing a workaround example for the time-being.

**Comment by graemerocher** [comment]
Looking at the code you stated `Currently, properties can be defaulted in application.* file, or in the package-info.java file as shown below.`

This is a actually not true, currently `@PropertySource` is not used anywhere apart from in MicronautTest and as a repeatable representation of `@Property`

So it would not be a case of updating `@PropertySource` since there is nothing to update.

Having said that the suggestions and ideas seem like good ones. I think it could technically be possible however there are some considerations. For example at the moment annotations on package configurations can in fact have `${..}` property references so there would have to be an initial pass through them to evaluate the `@PropertySource` definitions and add then to the environment loading any additional files.

So this is doable, just a matter of priorities. Of course if you want to see things advanced quicker PRs are of course welcome.

**Comment by graemerocher** [comment]
Triaged as Feature/enhancement. This remains a valid library-configuration ergonomics request, although @PropertySource is not currently a general-purpose file loader; the documented workaround is a custom PropertySourceLoader. Suggested fix: support explicit default resource names/files with deterministic precedence below application and environment-specific configuration, while defining how package-level property expressions are evaluated. Add tests showing library defaults overridden by application/env values.
